File: clients/ios/VisualPose/CoreJson.swift
import Foundation

enum CoreJson {
    static func poses(_ json: String) -> [CocoPose] {
        guard let root = Json.object(from: json), !hadError(root) else { return [] }
        guard let arr = Json.optArray(root, "poses") else { return [] }
        var out: [CocoPose] = []
        for item in arr {
            guard let poseObj = item as? [String: Any],
                  let kps = Json.optArray(poseObj, "keypoints")
            else { continue }
            var points: [CocoKeypoint] = []
            for kpItem in kps {
                guard let kp = kpItem as? [String: Any] else { continue }
                points.append(
                    CocoKeypoint(
                        x: Float(Json.optDouble(kp, "x", 0)),
                        y: Float(Json.optDouble(kp, "y", 0)),
                        confidence: Float(Json.optDouble(kp, "confidence", 0))
                    )
                )
            }
            if !points.isEmpty {
                out.append(CocoPose(keypoints: points))
            }
        }
        return out
    }

    static func bboxes(_ json: String) -> [BBox] {
        guard let root = Json.object(from: json), !hadError(root) else { return [] }
        guard let arr = Json.optArray(root, "detections") else { return [] }
        var out: [BBox] = []
        for item in arr {
            guard let node = item as? [String: Any],
                  let box = Json.optArray(node, "bbox_xyxy"),
                  box.count >= 4,
                  let x1 = Json.number(box[0]),
                  let y1 = Json.number(box[1]),
                  let x2 = Json.number(box[2]),
                  let y2 = Json.number(box[3])
            else { continue }
            out.append(BBox(x1: Float(x1), y1: Float(y1), x2: Float(x2), y2: Float(y2)))
        }
        return out
    }

    private static func hadError(_ root: [String: Any]) -> Bool {
        !Json.isNull(root, "error")
    }
}

File: clients/android/app/src/main/java/local/visual/corepose/CoreJson.kt
package local.visual.corepose

import org.json.JSONObject

object CoreJson {
    fun poses(json: String): List<CocoPose> {
        val root = JSONObject(json)
        if (!root.isNull("error")) {
            return emptyList()
        }
        val arr = root.optJSONArray("poses") ?: return emptyList()
        val out = ArrayList<CocoPose>(arr.length())
        for (i in 0 until arr.length()) {
            val poseObj = arr.optJSONObject(i) ?: continue
            val kps = poseObj.optJSONArray("keypoints") ?: continue
            val points = ArrayList<CocoKeypoint>(kps.length())
            for (k in 0 until kps.length()) {
                val kp = kps.optJSONObject(k) ?: continue
                points.add(
                    CocoKeypoint(
                        x = kp.optDouble("x", 0.0).toFloat(),
                        y = kp.optDouble("y", 0.0).toFloat(),
                        confidence = kp.optDouble("confidence", 0.0).toFloat(),
                    ),
                )
            }
            if (points.isNotEmpty()) {
                out.add(CocoPose(points))
            }
        }
        return out
    }

    fun bboxes(json: String): List<BBox> {
        val root = JSONObject(json)
        if (!root.isNull("error")) {
            return emptyList()
        }
        val arr = root.optJSONArray("detections") ?: return emptyList()
        val out = ArrayList<BBox>(arr.length())
        for (i in 0 until arr.length()) {
            val item = arr.optJSONObject(i) ?: continue
            val box = item.optJSONArray("bbox_xyxy") ?: continue
            if (box.length() < 4) {
                continue
            }
            out.add(
                BBox(
                    box.optDouble(0).toFloat(),
                    box.optDouble(1).toFloat(),
                    box.optDouble(2).toFloat(),
                    box.optDouble(3).toFloat(),
                ),
            )
        }
        return out
    }
}
